process/rating/peerHonesty/peerScore_test.go

```diff
@@ -0,0 +1,418 @@
+				return false
+			},
+			UpsertCalled: func(key string, span time.Duration) error {
+				upsertCalled = true
+				return nil
+			},
+		},
+		testscommon.NewCacherMock(),
+	)
+
+	pk := "pk"
+	topic := "topic"
+	units := int(cfg.MaxScore) + 1
+	pph.ChangeScore(pk, topic, units)
+
+	ps := pph.Get(pk)
+	assert.Equal(t, 1, len(ps.scores))
+	assert.Equal(t, cfg.MaxScore, ps.scores[topic])
+
+	assert.False(t, hasCalled)
+	assert.False(t, upsertCalled)
+}
+
+func TestP2pPeerHonesty_CheckBlacklistMinScoreReached(t *testing.T) {
+	t.Parallel()
+
+	cfg := createMockPeerHonestyConfig()
+	cfg.UnitValue = 4
+	hasCalled := false
+	upsertCalled := false
+	pph, _ := NewP2pPeerHonesty(
+		cfg,
+		&mock.TimeCacheStub{
+			HasCalled: func(key string) bool {
+				hasCalled = true
+				return false
+			},
+			UpsertCalled: func(key string, span time.Duration) error {
+				upsertCalled = true
+				return nil
+			},
+		},
+		testscommon.NewCacherMock(),
+	)
+
+	pk := "pk"
+	topic := "topic"
+	units := int(cfg.MinScore) - 1
+	pph.ChangeScore(pk, topic, units)
+
+	ps := pph.Get(pk)
+	assert.Equal(t, 1, len(ps.scores))
+	assert.Equal(t, cfg.MinScore, ps.scores[topic])
+
+	assert.True(t, hasCalled)
+	assert.True(t, upsertCalled)
+}
+
+func TestP2pPeerHonesty_CheckBlacklistHasShouldNotCallUpsert(t *testing.T) {
+	t.Parallel()
+
+	cfg := createMockPeerHonestyConfig()
+	cfg.UnitValue = 4
+	hasCalled := false
+	upsertCalled := false
+	pph, _ := NewP2pPeerHonesty(
+		cfg,
+		&mock.TimeCacheStub{
+			HasCalled: func(key string) bool {
+				hasCalled = true
+				return true
+			},
+			UpsertCalled: func(key string, span time.Duration) error {
+				upsertCalled = true
+				return nil
+			},
+		},
+		testscommon.NewCacherMock(),
+	)
+
+	pk := "pk"
+	topic := "topic"
+	units := int(cfg.MinScore) - 1
+	pph.ChangeScore(pk, topic, units)
+
+	assert.True(t, hasCalled)
+	assert.False(t, upsertCalled)
+}
+
+func TestP2pPeerHonesty_CheckBlacklistUpsertErrorsShouldWork(t *testing.T) {
+	t.Parallel()
+
+	cfg := createMockPeerHonestyConfig()
+	cfg.UnitValue = 4
+	upsertCalled := false
+	pph, _ := NewP2pPeerHonesty(
+		cfg,
+		&mock.TimeCacheStub{
+			HasCalled: func(key string) bool {
+				return false
+			},
+			UpsertCalled: func(key string, span time.Duration) error {
+				upsertCalled = true
+				return errors.New("expected error")
+			},
+		},
+		testscommon.NewCacherMock(),
+	)
+
+	pk := "pk"
+	topic := "topic"
+	units := int(cfg.MinScore) - 1
+	pph.ChangeScore(pk, topic, units)
+
+	assert.True(t, upsertCalled)
+}
+
+func TestP2pPeerHonesty_ApplyDecay(t *testing.T) {
+	t.Parallel()
+
+	cfg := createMockPeerHonestyConfig()
+	pph, _ := NewP2pPeerHonesty(
+		cfg,
+		&mock.TimeCacheStub{},
+		testscommon.NewCacherMock(),
+	)
+
+	pks := []string{"pkMin", "pkMax", "pkNearZero", "pkZero", "pkValue"}
+	initial := []float64{cfg.MinScore, cfg.MaxScore, approximateZero / 2, 0, 28}
+	topic := "topic"
+
+	for idx := range pks {
+		pph.Put(pks[idx], topic, initial[idx])
+	}
+
+	pph.applyDecay()
```

Comment:
Perhaps add an extra test that calls `applyDecay` a lot of times, just to assert that all scores go to zero?

Reply:
done